p2p/discovery/mdns.go

```diff
@@ -37,186 +36,183 @@ type Notifee interface {
+	s := &mdnsService{
+		ctx:             ctx,
+		ctxCancel:       cancel,
+		resolverRunning: make(chan struct{}),
+		host:            host,
+		serviceName:     serviceName,
+	}
+	s.startServer()
+	s.startResolver()
+	return s
+}
+
+func (s *mdnsService) Close() error {
+	s.ctxCancel()
+	if s.server != nil {
+		s.server.Shutdown()
+	}
+	<-s.resolverRunning
+	return nil
+}
+
+// We don't really care about the IP addresses, but the spec (and various routers / firewalls) require us
+// to send A and AAAA records.
+func (s *mdnsService) getIPs(addrs []ma.Multiaddr) ([]string, error) {
+	var ip4, ip6 string
 	for _, addr := range addrs {
-		na, err := manet.ToNetAddr(addr)
+		network, hostport, err := manet.DialArgs(addr)
 		if err != nil {
 			continue
 		}
-		tcp, ok := na.(*net.TCPAddr)
-		if ok {
-			out = append(out, tcp)
+		host, _, err := net.SplitHostPort(hostport)
+		if err != nil {
+			continue
 		}
+		if ip4 == "" && (network == "udp4" || network == "tcp4") {
+			ip4 = host
+		} else if ip6 == "" && (network == "udp6" || network == "tcp6") {
+			ip6 = host
+		}
+	}
+	ips := make([]string, 0, 2)
+	if ip4 != "" {
+		ips = append(ips, ip4)
 	}
-	if len(out) == 0 {
-		return nil, errors.New("failed to find good external addr from peerhost")
+	if ip6 != "" {
+		ips = append(ips, ip6)
 	}
-	return out, nil
+	if len(ips) == 0 {
+		return nil, errors.New("didn't find any IP addresses")
+	}
+	return ips, nil
 }
 
-func NewMdnsService(ctx context.Context, peerhost host.Host, interval time.Duration, serviceTag string) (Service, error) {
-
-	var ipaddrs []net.IP
-	port := 4001
+func (s *mdnsService) mdnsInstance() string {
+	return string(s.host.ID())
+}
 
-	addrs, err := getDialableListenAddrs(peerhost)
+func (s *mdnsService) startServer() error {
+	interfaceAddrs, err := s.host.Network().InterfaceListenAddresses()
 	if err != nil {
-		log.Warn(err)
-	} else {
-		port = addrs[0].Port
-		for _, a := range addrs {
-			ipaddrs = append(ipaddrs, a.IP)
-		}
+		return err
 	}
-
-	myid := peerhost.ID().Pretty()
-
-	info := []string{myid}
-	if serviceTag == "" {
-		serviceTag = ServiceTag
-	}
-	service, err := mdns.NewMDNSService(myid, serviceTag, "", "", port, ipaddrs, info)
+	addrs, err := peer.AddrInfoToP2pAddrs(&peer.AddrInfo{
+		ID:    s.host.ID(),
+		Addrs: interfaceAddrs,
+	})
 	if err != nil {
-		return nil, err
+		return err
 	}
-
-	// Create the mDNS server, defer shutdown
-	server, err := mdns.NewServer(&mdns.Config{Zone: service})
-	if err != nil {
-		return nil, err
+	var txts []string
+	for _, addr := range addrs {
+		if manet.IsThinWaist(addr) { // don't announce circuit addresses
+			txts = append(txts, dnsaddrPrefix+addr.String())
+		}
 	}
 
-	s := &mdnsService{
-		server:   server,
-		service:  service,
-		host:     peerhost,
-		interval: interval,
-		tag:      serviceTag,
+	ips, err := s.getIPs(addrs)
+	if err != nil {
+		return err
+	}
+
+	server, err := zeroconf.RegisterProxy(
+		s.mdnsInstance(),
+		s.serviceName,
+		mdnsDomain,
+		4001,
```

Comment:
Can we leave a comment saying "this port doesn't actually matter, only the TXT records are used?"

Reply:
Does this need to be lower case?

Reply:
Done. Not sure what you're referring to when you ask about lower case.

Reply:
Ah, sorry, commented on the wrong line. I was referring to the peer ID in the line below.